mne/beamformer/_rap_music.py

```diff
@@ -0,0 +1,248 @@
+    explained_data : array
+        Data explained by the dipoles using a least square fitting with the
+        selected active dipoles and their estimated orientation.
+        Computed only if return_explained_data is True.
+    """
+    is_free_ori, ch_names, proj, vertno, G = _prepare_beamformer_input(
+        info, forward, label=None, picks=picks, pick_ori=None)
+
+    gain = G.copy()
+
+    # Handle whitening + data covariance
+    whitener, _ = compute_whitener(noise_cov, info, picks)
+    if info['projs']:
+        whitener = np.dot(whitener, proj)
+
+    # whiten the leadfield and the data
+    G = np.dot(whitener, G)
+    data = np.dot(whitener, data)
+
+    eig_values, eig_vectors = linalg.eigh(np.dot(data, data.T))
+    phi_sig = eig_vectors[:, -signal_ndim:]
+
+    n_orient = 3 if is_free_ori else 1
+    A = np.zeros((G.shape[0], n_dipoles))
+    active_set = []
+    oris = []
+
+    G_proj = G
+    phi_sig_proj = phi_sig
+
+    for k in range(n_dipoles):
+        subcorr_max = -1.
+        source_idx = None
+        for i_source in range(G.shape[1] // n_orient):
+            Gk = G_proj[:, n_orient * i_source:
+                        n_orient * i_source + n_orient]
+
+            subcorr, ori = _compute_subcorr(Gk, phi_sig_proj)
+            if ori[-1] < 0:  # make sure ori is relative to surface ori
+                ori *= -1.
+            if subcorr > subcorr_max:
+                subcorr_max = subcorr
+                source_idx = i_source
+                source_ori = ori
+                A[:, k] = np.dot(Gk, ori)
+                if n_orient == 1:
+                    ori = [forward['src'][0]['nn'][vertno[0][i_source]]
+                           if i_source <= vertno[0].size else
+                           forward['src'][1]['nn'][vertno[1][i_source -
+                                                             vertno[0].size]]]
+
+        active_set.append(source_idx)
+        oris.append(source_ori)
+
+        logger.info("source %s found: p = %s" % (k + 1, active_set[-1]))
+        if n_orient == 3:
+            logger.info("ori = %s %s %s" % (source_ori[0], source_ori[1],
+                                            source_ori[2]))
+
+        projection = _compute_proj(A[:, :k + 1])
+        G_proj = np.dot(projection, G)
+        phi_sig_proj = np.dot(projection, phi_sig)
+
+    sol = linalg.lstsq(A, data)[0]
+
+    active_set = np.sort(active_set)
+    explained_data = None
+    if return_explained_data:
+        explained_data = np.dot(gain[:, active_set], sol)
+
+    vertno[1] = vertno[1][active_set[active_set > vertno[0].size] -
+                          vertno[0].size]
+    vertno[0] = vertno[0][active_set[active_set <= vertno[0].size]]
+
+    tstep = 1.0 / info['sfreq']
+
+    return _make_dipoles(times, tstep, forward['src'], vertno, active_set,
+                         oris, sol), explained_data
+
+
+def _make_dipoles(times, tstep, src, vertno, active_set, oris, sol):
+    times *= 1000
```

Comment:
this is dangerous you do inplace modification in a function. So edge effects... is it on purpose?